Fix deployment list

Mahmoud-Emad opened the pull request:
### Description

There is an issue with listing VMs due to the merge of the Caprover PR

### Changes

- Keep the worker listed if the user deletes only the leader.
- Remove the 'Manage Workers' button if the listed VM is a worker.
- Remove the 'View Admin Panel' button if the listed VM is a worker.
- Check if the full deployment has a 'leader' instead.
- Fix listing issues in the other solutions.
- Update the mergeCaproverDeployment function.
- Fix an issue with adding a worker on a deployed Caprover leader

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3542

### Tested Scenarios

A list of scenarios tried to match the deliverables

### To consider

#### Preliminary Checks:

- Does it completely address the issue linked?
- What about edge cases?
- Does it meet the specified acceptance criteria?
- Are there any unintended side effects?
- Does the PR adhere to the team's coding conventions, style guides, and best practices?
- Does it integrate well with existing features?
- Does it impact the overall performance of the application?
- Are there any bottlenecks or slowdowns?
- Has it been optimized for efficiency?
- Has it been adequately tested with unit, integration, and end-to-end tests?
- Are there any known defects or issues?
- Is the code well-documented?
- Are changes to documentation reflected in the code?

#### UI Checks:

- If a UI design is provided/ does it follow it?
- Does every button work?
- Is the data displayed logical? Is it what you expected?
- Does every validation work?
- Does this interface feel intuitive?
- What about slow network? Offline?
- What if the gridproxy/graphql/chain is failing?
- What would a first time user have a hard time navigating here?

#### Code Quality Checks:

- Code formatted/linted? Are there unused imports? .. etc
- Is there redundant/repeated code?
- Are there conditionals that are always true or always false?
- Can we write this more concisely?
- Can we reuse this code? If yes, where?
- Will the changes be easy to maintain and update in the future?
- Can this code become too complex to understand for other devs?
- Can this code cause future integration problems?

### Checklist

- [ ] Tests included
- [x] Build pass
- [ ] Documentation
- [x] Code format and docstrings
- [ ] Screenshots/Video attached (needed for UI changes)

Mahmoud-Emad committed:
fix: Fix deployment list:

- Keep the worker listed if the user deletes only the leader.
- Remove the 'Manage Workers' button if the listed VM is a worker.
- Remove the 'View Admin Panel' button if the listed VM is a worker.
- Check if the full deployment has a 'leader' instead.
- Fix listing issue in the other solutions.
- Update the mergeCaproverDeployment function.
- Fix an issue with adding a worker on a deployed Caprover leader

Mahmoud-Emad committed:
Merge branch 'development' into development_fix_listing

Mahmoud-Emad commented:
### Tested scenarios

- Listing in the Micro/Full VMs solutions
- Listing a valid Caprover cluster
- Listing an invalid Caprover cluster <workers only> see the attached image

![image](https://github.com/user-attachments/assets/94b0d545-62cd-444a-be94-3fc1d1be704e)

samaradel commented on packages/playground/src/utils/deploy_helpers.ts:
```diff
@@ -11,29 +13,46 @@ export function createNetwork(network: Network = {}): NetworkModel {
-export function setCaproverWorkers(vms: ZmachineData[], projectName: string | undefined = undefined): ZmachineData {
-  let leader: any = null;
-  const workers: any[] = [];
+export const mergeCaproverDeployments = (clusters: ZmachineData[][], projectName: string) => {
```
I see that ```projectName``` not used in the function anymore

Mahmoud-Emad replied:
indeed, removed

amiraabouhadid committed:
fix max name length

amiraabouhadid committed:
Revert "fix max name length"

This reverts commit f433a884ec9270b62c7faec5b8926a94011fc2df.

Mahmoud-Emad committed:
fix: Remove the projectName from function mergeCaproverDeployments parameters

Mahmoud-Emad committed:
Merge branch 'development' into development_fix_listing

Mahmoud-Emad commented:
Converted to draft due to an issue happened when deleting the instances

Mahmoud-Emad committed:
fix: Prevent race condition in deployment deletion:

- Replaced `Promise.all` with a `.reduce()` approach to sequentially await each `deleteDeployment` and `deleteGatewayDeployment` call.

Mahmoud-Emad commented:
> Converted to draft due to an issue happened when deleting the instances

Should be fixed now

### Investigation and Solution:

This issue might be due to how `Promise.all` handles errors versus the sequential `for` loop. With `Promise.all`, all `deleteDeployment` calls are executed concurrently, meaning the `delete_machine` method might check for the deployment existence before another instance of it finishes deleting. In contrast, a regular `for` loop executes sequentially, so each `deleteDeployment` call completes before the next starts, avoiding potential race conditions.


### Work Completed:

**Prevent a race condition in deployment deletion by switching to sequential execution**

- Replaced `Promise.all` with a `.reduce()` approach to sequentially await each `deleteDeployment` and `deleteGatewayDeployment` call.

0oM4R reviewed (commented):
![Screenshot from 2024-10-27 12-03-04](https://github.com/user-attachments/assets/3772c62d-3a5e-4292-8ecd-a2c408ece0e3)
![Screenshot from 2024-10-27 12-02-25](https://github.com/user-attachments/assets/9f88541e-78bd-4e4d-a5ea-7f3d85eb5f8a)


caprover deployments didn't show up on the deployment list, they were listed but after click on delete they are still on the contracts list,

samaradel commented:
I faced the same delete issue again
![image](https://github.com/user-attachments/assets/23560c36-f622-408f-a095-c7b3e6592276)

Mahmoud-Emad committed:
Merge branch 'development' into development_fix_listing

Mahmoud-Emad committed:
fix: Fix deleting the recently added worker

0oM4R commented:
> ![Screenshot from 2024-10-27 12-03-04](https://private-user-images.githubusercontent.com/62248851/380482520-3772c62d-3a5e-4292-8ecd-a2c408ece0e3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxOTI4ODIsIm5iZiI6MTczMDE5MjU4MiwicGF0aCI6Ii82MjI0ODg1MS8zODA0ODI1MjAtMzc3MmM2MmQtM2E1ZS00MjkyLThlY2QtYTJjNDA4ZWNlMGUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI5VDA5MDMwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI3ZDA2OTBkOTRiNWQ3ZWRiYjFjN2FjOWI3ZmI2YjQxNjU2MzVlZjAxOThmZThjYTZiMDM0ZTVmYThkNDI2ODkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.z9WKkghvAuOqOp7pLceufholOyLw9e_1KysjTGSQD4o) ![Screenshot from 2024-10-27 12-02-25](https://private-user-images.githubusercontent.com/62248851/380482525-9f88541e-78bd-4e4d-a5ea-7f3d85eb5f8a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxOTI4ODIsIm5iZiI6MTczMDE5MjU4MiwicGF0aCI6Ii82MjI0ODg1MS8zODA0ODI1MjUtOWY4ODU0MWUtNzhiZC00ZTRkLWE1ZWEtN2YzZDg1ZWI1ZjhhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI5VDA5MDMwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg4MjJiODMyMTI1YWVhMDRjYjIwMmMxZGExYWNiZjlhYTM4MTI0NmMxZDZjYjI4ZDdmNDBkYjZlYmYxMDBhYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.BrvbDR2D6_oiG7qvYXe4UWcGbOwNdPvGf-SFswS0moI)
> 
> caprover deployments didn't show up on the deployment list, they were listed but after click on delete they are still on the contracts list,

still happening.. 
 
![Screenshot from 2024-10-29 11-59-38](https://github.com/user-attachments/assets/5eb100b5-4527-4f23-9fe0-50a017c1ae95)

 
also while deleting the cluster it shows 5 chips while i only have 1 leader and 2 workers 
<img width="640" alt="image" src="https://github.com/user-attachments/assets/e9f85424-6649-46f5-9037-0bac9a651cc5">
![Screenshot from 2024-10-29 11-45-37](https://github.com/user-attachments/assets/e973e67e-d891-4381-b137-0f2db2b402a2)


### will ignore the caprover for now

0oM4R reviewed (approved):
listing deployments works fine. 
will reopen caprover issue
and will open new issue to revisit deployment list changes
